src/Core/Component.js

```diff
@@ -118,223 +147,295 @@ function mountComponentElement (element) {
 
 	if (owner[SharedComponentDidUpdate])
 		getLifecycleUpdate(element, SharedComponentDidUpdate, prevProps, prevState, nextContext)
 
 	if (element.ref !== snapshot.ref)
-		commitRefs(element, snapshot.ref, SharedReferenceReplace)
-
-	element.work = SharedWorkIdle
+		commitRefs(element, snapshot.ref, SharedRefsReplace)
 }
 
 /**
  * @param {Element} element
  */
 function unmountComponentElement (element) {
-	if ((element.state = null, element.owner[SharedComponentWillUnmount]))
-		element.state = getLifecycleMount(element, SharedComponentWillUnmount)
+	if (element.owner[SharedComponentWillUnmount])
+		if (element.cache = getLifecycleUnmount(element, SharedComponentWillUnmount, element.owner))
+			if (thenable(element.cache))
+				return
+
+	element.cache = null
 }
 
 /**
  * @param {Element} element
- * @param {Component} instance
- * @param {function?} callback
+ * @param {Element} snapshot
  * @param {number} signature
+ * @return {function}
  */
-function enqueueComponentUpdate (element, instance, callback, signature) {
-	if (!element)
-		return void requestAnimationFrame(function () {
-			enqueueComponentUpdate(getComponentElement(instance), instance, callback, signature)
-		})
-
-	if (element.work === SharedWorkProcessing)
-		return void requestAnimationFrame(function () {
-			enqueueComponentUpdate(element, instance, callback, signature)
-		})
-
-	if (element.active)
-		updateComponent(element, element, signature)
-	else
-		instance.state = assign({}, instance.state, element.state)
+function enqueueComponentElement (element, snapshot, signature) {
+	return function then () {
+		updateComponentElement(element, snapshot, signature)
+	}
+}
 
-	if (callback)
-		enqueueStateCallback(element, instance, callback)
+/**
+ * @param {Element} element
+ * @param {Component} owner
+ * @param {object} state
+ * @return {function}
+ */
+function enqueueComponentState (element, owner, state) {
+	return function then (resolve, reject) {
+		enqueueStatePromise(element, owner, state).then(function () {
+			resolve(element.children.type.then === then && getComponentSnapshot(element, owner))
+		}, reject)
+	}
 }
 
 /**
  * @param {Element} element
- * @param {Element} snapshot
- * @param {number} signature
+ * @param {AsyncGenerator} generator
+ * @param {object} cache
+ * @return {object}
  */
-function enqueuePendingUpdate (element, snapshot, signature) {
-	return function () {
-		updateComponent(element, snapshot, signature)
+function enqueueComponentGenerator (element, generator, cache) {
+	return function then (resolve, reject) {
+		requestAnimationFrame(function () {
+			generator.next(cache.value).then(function (value) {
+				if (value.done === true && value.value === undefined)
+					return !element.active && resolve(getElementDefinition(cache.value))
+
+				if ((cache.value = value.value, element.active))
+					resolve(getElementDescription(value.value))
+
+				then(resolve, reject)
```

Comment:
This was intentionally added to support patterns like

```js
async function * Foo ({value}) {
	var step = 0
	
	while (true) {
		if (step)
			yield h('h1', 'Hello ', name)
		else
			yield h('form', {onSubmit: (e) => e.preventDefault(step++)},
				h('input', {placeholder: 'Name?', onInput: ({target}) => name = target.value})
			)
	}
}
```

Without the need to wrap it a Promise, i.e

```js
async function * Foo ({value}) {
	var step = 0
	
	while (true) {
		yield new Promise((resolve) => {
			if (step)
				resolve(h('h1', 'Hello ', name))
			else
				resolve(h('form', {onSubmit: (e) => e.preventDefault(step++)},
					h('input', {placeholder: 'Name?', onInput: ({target}) => name = target.value})
				))
		})
	}
}
```

But i agree this probably shouldn't be at the library level since the pattern you describe is only possible outside of the `requestAnimationFrame` heuristic.

Reply:
Doesn't looping `requestAnimationFrame` all the time have a big hit on performance?

Reply:
Another thing is when `Foo` props change and it re-render, what happens?

Reply:
Yes, i think we can remove requestAnimationFrame, the example is interesting for animation patterns but still possible without the use of requestAnimationFrame within the implementation details. 

>Another thing is when Foo props change and it re-render, what happens?

Given the following

```js
async function * Foo ({children, time}) {
	yield wait(h('h1', 'Loading'), time)
	yield h('h1', children)
}

render(h(Foo, {children: 'Hello', time: 2000}), container)
render(h(Foo, {children: 'World', time: 1000}), container)
```

I think it should be spec'ed to render the last value `World` vs the last resolved value `Hello`, what do you think?

Reply:
agreed